### PR TITLE
Improve beamformer performance with >32 dual-pol beams

### PR DESCRIPTION
**Pull request by bmerry**
Checklist (if not applicable, edit to add `(N/A)` and mark as done):

- [x] (N/A) If dependencies are added/removed: update `setup.cfg` and `.pre-commit-config.yaml`
- [x] (N/A) If modules are added/removed: use `sphinx-apidoc -efo doc/ src/` to update files in `doc/`
- [x] Ensure copyright notices are present and up-to-date
- [x] (N/A) If qualification tests are changed: attach a sample qualification report
- [x] If design has changed: ensure documentation is up to date
- [x] (N/A) If ICD-defined sensors have been added: update `fake_servers.py` in katsdpcontroller to match

### CODE REVIEW

File: doc/xbgpu.design.b.rst [hunk 1]
```diff
@@ -52,9 +52,19 @@ sizes have two advantages:
...
    the coefficient computation less efficient.
 
-The design above works well for small numbers of beams, but the register usage
-scales with the number of beams. Supporting significantly more beams would
-require rework to process beams in smaller batches.
+Higher beam counts
+^^^^^^^^^^^^^^^^^^
+The design above works well for small numbers of beams (up to about 64
+single-pol beams), but the register usage scales with the number of beams and
+eventually the registers spill to memory, causing very poor performance.
```
Comment (amishatishpatel) [inline doc/xbgpu.design.b.rst:59]: I have to clarify, Re:
> and eventually the registers spill to memory

What/which memory are you referring to here? It's "all GPU memory", so I would think it spills over to System memory? (Admittedly not entirely familiar with the term/phrase)

File: src/katgpucbf/xbgpu/kernels/beamform.mako [hunk 2]
```diff
@@ -88,64 +94,78 @@ KERNEL REQD_WORK_GROUP_SIZE(BLOCK_SPECTRA, 1, 1) void beamform(
...
+                    }
                 }
             }
+            BARRIER();  // protect against next loop iteration overwriting l_weights
         }
-        BARRIER();  // protect against next loop iteration overwriting l_weights
-    }
-
-    if (!valid)
-        return;
```
Comment (amishatishpatel) [inline src/katgpucbf/xbgpu/kernels/beamform.mako:140]: The removal of this branch of code makes me wonder what it was used for in the first place - seeing as it was checked so late in the day/kernel?

**Review by amishatishpatel — COMMENTED** [review]
Really neat, really cool update. Just a couple questions before I can explicitly Approve.

Also, I'm still unsure whether this requires an update to `beamform.py`
* The writing to `out` over the batches still happens for all `b_batch`es unrolled,
* I don't think the way the kernel is invoked needs to change either (I'm not too familiar with mako syntax anyway to really critique)
  * I gather it will stil acknowledge your passing of `"beam_pols": self.beam_pols` as a recognised parameter - just for a different use now.

File: doc/xbgpu.design.b.rst (hunk 1 above)
Comment (james-smith-za) [reply doc/xbgpu.design.b.rst:59]: Registers are the little boxes very close to the CPU which some working data is stored in.

If you have too many variables then data needs to get stored in global memory which has quite a long round-trip if you need to read it again.

**Review by amishatishpatel — APPROVED** [review]

**Pull request merged**